gapic-generator-java/src/main/java/com/google/api/generator/gapic/composer/defaultvalue/DefaultValueComposer.java

```diff
@@ -442,6 +449,87 @@ public static Expr createSimpleMessageBuilderValue(
+   * try and generate the default value for the configured resource reference. This allows a
+   * fallback option where if this check fails, then the generator can attempt to generate a default
+   * value using {@link #createValue(Field, boolean, Map, Map, Map, HttpBindings)}. Additionally,
+   * there are references to createResourceHelperValue being used elsewhere (client header samples,
+   * sample code gen) which is a much larger effort to change.
+   *
+   * <p>Rationale for matching the Resource's pattern with HttpBindings: If the field is used in the
+   * path, check that the field's resource reference matches the RPC's HttpBindings. There is an
+   * edge case where an RPC's HttpBindings defined in the proto may not match a message. This is
+   * possible for Mixin RPCs if service teams configures a custom HttpBindings in the service yaml
+   * file. For example, the GetIamPolicyRequest's resource field has a resource reference to match
+   * anything (`*`). This normally works because the GetIamPolicy RPC will take in any resource
+   * (HttpBindings is defined with `{resource=**}`). But if the custom path is expected to take in a
+   * resource with a more specific path (i.e. `{resource=service/*}), it won't work unless the
+   * resource has a `service/*` pattern. This logic will try to match a resource that has the
+   * pattern `service/*`.
+   */
+  static List<ResourceName> getMatchingResources(
+      Field field,
+      Map<String, ResourceName> resourceNames,
+      HttpBindings bindings,
+      Map<String, String> valuePatterns) {
+    if (!field.hasResourceReference()) {
+      return ImmutableList.of();
+    }
+    String resourceTypeString = field.resourceReference().resourceTypeString();
+    if (!resourceNames.containsKey(resourceTypeString)) {
+      return ImmutableList.of();
+    }
+    ResourceName resourceReference = resourceNames.get(resourceTypeString);
+
+    // Resource reference (google.api.resource_reference) is a string type with a field name of
+    // `parent` or `name`. It won't have anything nested in the bindings (i.e. `{name.field=*}`).
+    // This check is needed as a message can contain fields with valid resource references that
+    // aren't used in the path. Additionally, this flag only impacts HttpJson callables as gRPC
+    // does not need to match the request the callable's URL. This method is used to generate
+    // default values for gRPC callables as well but gRPC's valuePatterns map will be empty and
+    // the logic will be skipped.
+    boolean isFieldUsedInPath = bindings != null && valuePatterns.containsKey(field.name());
+
+    // Holds the list of known resources (excluding the wildcard resource).
+    List<ResourceName> possibleResources =
+        resourceNames.values().stream()
+            .filter(x -> !x.isOnlyWildcard())
+            .collect(Collectors.toList());
+
+    if (resourceReference.isOnlyWildcard()) {
+      if (isFieldUsedInPath) {
```

Comment:
Why not combine these two if-statements?

Reply:
There are *two* possible paths if resourceReference is wildcard:
- If used in path, then filter for the pattern that matches the bindings
- If not used in the path, then return all possible resources